### PR TITLE
[cf-core, node, client] Unrolled deposit

### LINKED ISSUE #462: [protocol, card, client] Unrolled Deposit
Separate out the deposit commitment from the onchain deposit transaction itself.

WIP: https://github.com/counterfactual/monorepo/issues/2516

### PR DESCRIPTION
**Pull request by rhlsthrm**
## The Problem
Fixes #462 

## Checklist:

- [x] I am making this PR against staging, not master
- [x] My code follows the code style of this project.
- [ ] I have described any backwards-incompatibility implications above.
- [ ] I have highlighted which parts of the code should be reviewed most carefully.
- [ ] If my change requires a change to the documentation, I have updated it accordingly.

### CODE REVIEW

File: modules/cf-core/src/api.ts [hunk 1]
```diff
@@ -35,11 +37,13 @@ const controllers = [
   CreateChannelController,
   DepositController,
   InstallAppInstanceController,
+  InstallBalanceRefundController,
```
Comment (snario) [inline modules/cf-core/src/api.ts]: A better name might be "RequestDepositRightsController" or something more descriptive of what the intent is.

File: modules/cf-core/src/machine/types.ts [hunk 2]
```diff
@@ -119,6 +119,7 @@ export type UninstallParams = {
   initiatorXpub: string;
   responderXpub: string;
   multisigAddress: string;
+  blockNumberToUseIfNecessary?: number;
```
Comment (snario) [inline modules/cf-core/src/machine/types.ts]: `ToUseIfNecessary` might be better described as `AtWhichUninstallIsCalled`

File: modules/cf-core/src/api.ts (hunk 1 above)
Comment (rhlsthrm) [reply modules/cf-core/src/api.ts]: Thanks, this is a great name.

File: modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        ...params,
+        // unused params to make types happy
+        tokenAddress: CONVENTION_FOR_ETH_TOKEN_ADDRESS,
+        amount: Zero
+      },
+      blockNumber
+    );
+
+    return {
+      multisigBalance
```
Comment (snario) [inline modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts]: Why is this the return type?

File: modules/cf-core/src/protocol/utils/get-outcome-increments.ts [hunk 4]
```diff
@@ -84,14 +86,19 @@ export async function computeTokenIndexedFreeBalanceIncrements(
...
-    if (amount.gt(0)) {
+    const currentBlockNumber = await provider.getBlockNumber();
+    if (
+      amount.gt(0) ||
+      (blockNumberToUseIfNecessary &&
+        currentBlockNumber >= blockNumberToUseIfNecessary)
+    ) {
       return {
         [(appInstance.state as CoinBalanceRefundState).tokenAddress]: {
           [to]: amount
```
Comment (snario) [inline modules/cf-core/src/protocol/utils/get-outcome-increments.ts]: I think this is right, but it would benefit from declaring some variables with names, or adding some comments, as it reads confusingly.

File: modules/cf-types/.prettierrc [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+  "printWidth": 80,
+  "trailingComma": "none"
+}
```
Comment (snario) [inline modules/cf-types/.prettierrc]: Seems unrelated to this PR.

Comment (LayneHaber) [reply modules/cf-types/.prettierrc]: it was to fix the linting from when we added these packages into our repo, but yes otherwise unrelated to this PR

File: modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts (hunk 3 above)
Comment (rhlsthrm) [reply modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts]: Callers might be interested in the multisig balance after the balance refund app is installed. Open to other suggestions on return values.

File: modules/cf-types/.prettierrc (hunk 5 above)
Comment (rhlsthrm) [reply modules/cf-types/.prettierrc]: Correct. This has been cherry-picked into other branches as well. Will leave it for simplicity.

File: modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts (hunk 3 above)
Comment (LayneHaber) [reply modules/cf-core/src/methods/state-channel/uninstall-balance-refund/controller.ts]: Agree -- should be at least something closer to the updated free balance?

File: modules/client/src/connext.ts [hunk 6]
```diff
@@ -227,6 +230,30 @@ export const connect = async (opts: ClientOptions): Promise<IConnextClient> => {
   log.debug("Registering subscriptions");
   await client.registerSubscriptions();
 
+  // TODO: should we have a flag for this?
```
Comment (LayneHaber) [inline modules/client/src/connext.ts]: as in a flag as to whether this app is already installed or an event thrown or what

File: modules/client/src/connext.ts [hunk 7]
```diff
@@ -235,7 +262,7 @@ export const connect = async (opts: ClientOptions): Promise<IConnextClient> => {
   // since if the hub never submits you should not continue interacting
   log.debug("Reclaiming pending async transfers");
   // no need to await this if it needs collateral
-  client.reclaimPendingAsyncTransfers();
+  await client.reclaimPendingAsyncTransfers();
```
Comment (LayneHaber) [inline modules/client/src/connext.ts]: if it needs collateral why would we await these reclaims? would take forever for the client to start up

File: modules/client/src/connext.ts (hunk 6 above)
Comment (rhlsthrm) [reply modules/client/src/connext.ts]: A flag to enable async deposits to a token.

File: modules/client/src/connext.ts (hunk 7 above)
Comment (rhlsthrm) [reply modules/client/src/connext.ts]: @bohendo found out this was causing issues, but now I'm realizing maybe we shouldn't do this. Although maybe we should wait so that we don't let things get weird if you try to do lots of other things while this is happening?
